core/src/main/scala/kafka/server/TokenManager.scala

```diff
@@ -0,0 +1,528 @@
+      ScramCredentialUtils.credentialToString(credential)
+    }
+
+    for (mechanism <- ScramMechanism.values)
+      scramConfig.setProperty(mechanism.mechanismName, scramCredential(mechanism))
+  }
+
+  /**
+    *
+    * @param owner
+    * @param renewers
+    * @param maxTimePeriodMs
+    * @param responseCallback
+    */
+  def createToken(owner: KafkaPrincipal,
+                  renewers: List[KafkaPrincipal],
+                  maxTimePeriodMs: Long,
+                  responseCallback: createResponseCallback) {
+
+    if (!config.tokenAuthEnabled) {
+      responseCallback(CreateTokenResult(-1, -1, -1, "", Array[Byte](), Errors.TOKEN_AUTH_DISABLED))
+    } else {
+
+      //Can we use UUID here?
+      val sequenceNum: Integer = zkUtils.getSequenceId(ZkUtils.TokenSequenceIdPath)
+      val tokenId = TokenSequenceIdPrefix + sequenceNum
+
+      val issueTime = time.milliseconds
+      val maxLifeTime = if (maxTimePeriodMs <= 0) tokenMaxLifetime else maxTimePeriodMs
+      val maxLifeTimeStamp = issueTime + maxLifeTime
+      val expiryTime = Math.min(maxLifeTimeStamp, issueTime + tokenRenewInterval)
+
+      val tokenInfo = new TokenInformation(tokenId, owner, renewers.asJava)
+      tokenInfo.setIssueDate(issueTime)
+      tokenInfo.setMaxDate(maxLifeTimeStamp)
+      tokenInfo.setExpiryDate(expiryTime)
+
+      val password = createPassword(tokenId, secretKey)
+
+      try {
+        val token = new DelegationToken(tokenInfo, password)
+        updateToken(tokenId, token)
+        info(s"Created a delegation token : $tokenId for owner : $owner")
+        responseCallback(CreateTokenResult(issueTime, expiryTime, maxLifeTimeStamp, tokenId, password, Errors.NONE))
+      } catch {
+        case e: Throwable => responseCallback(CreateTokenResult(-1, -1, -1, "", Array[Byte](), Errors.forException(e)))
+      }
+    }
+
+  }
+
+  /**
+    *
+    * @param principal
+    * @param hmac
+    * @param renewTimePeriodMs
+    * @param renewCallback
+    */
+  def renewToken(principal: KafkaPrincipal,
+                 hmac: ByteBuffer,
+                 renewTimePeriodMs: Long,
+                 renewCallback: renewResponseCallback) {
+
+    if (!config.tokenAuthEnabled) {
+      renewCallback(Errors.TOKEN_AUTH_DISABLED, -1)
+    } else {
+      lock.synchronized  {
+        getToken(hmac) match {
+          case Some(token) => {
+            val now = time.milliseconds
+            val tokenInfo =  token.tokenInfo
+
+            if (!allowed(principal, tokenInfo)) {
+              renewCallback(Errors.TOKEN_OWNER_MISMATCH, -1)
+            } else if (tokenInfo.maxDate < now || tokenInfo.expiryDate < now) {
+              renewCallback(Errors.TOKEN_EXPIRED, -1)
+            } else {
+              val renewTimePeriod = if (renewTimePeriodMs < 0) tokenRenewInterval else renewTimePeriodMs
+              val renewTimeStamp = now + renewTimePeriod
+              val expiryTime = Math.min(tokenInfo.maxDate, renewTimeStamp)
```

Comment:
I think both cases could be argued here. On the one hand it has the potential for "fraud", if for example a disgruntled employee leaves the company, briefly changes config.delegationTokenMaxLifeMs to something very large and creates a token for him to use later on..
On the other hand, having changes to this parameter effect existing tickets has the potential to create a bit of havoc on large deployments when all of a sudden tokens stop working that the users planned to get recreated sometime down the line. I've seen a lot of production issues caused by expired ssl certificates - and the expiration date on those is fixed. 

I think my vote would be to not change token expiration dates when updating delegationTokenMaxLifeMs, since the end dates will be different based on creation time anyway. Perhaps it is an option to create an admin tool to display tokens that exceed the currently configured maxtime?